src/lib/reasoners/uf.ml

```diff
@@ -999,100 +999,206 @@ let assign_next env =
+  let { E.f; xs; ty; _ } = E.term_view t in
+  if is_forbidden_symbol f ty
+  || E.is_internal_name t || E.is_internal_skolem t
+  || E.equal t E.vrai || E.equal t E.faux
   then
     acc
   else
-    let xs, tys, mrepr =
-      List.fold_left
-        (fun (xs, tys, mrepr) x ->
-           let rep_x, mrepr = model_repr_of_term x env mrepr None in
-           assert (is_const_term rep_x);
-           (x, rep_x)::xs,
-           (E.type_info x)::tys,
-           mrepr
-        ) ([],[], mrepr) (List.rev xs)
-    in
-    let rep, mrepr = model_repr_of_term t env mrepr unbounded in
-    assert (is_a_good_model_value rep);
-    assert (is_const_term rep);
-    match f, xs, ty with
-    | Sy.Op Sy.Set, _, _ -> acc
-
-    | Sy.Op Sy.Get, [(_,(a,_));((_,(i,_)) as e)], _ ->
-      begin
-        match X.term_extract a with
-        | Some ta, true ->
-          let { E.f = f_ta; xs=xs_ta; _ } = E.term_view ta in
-          assert (xs_ta == []);
-          fprofs,
-          cprofs,
-          ModelMap.add (f_ta,[X.type_info i], ty) ([e], rep) carrays,
-          mrepr
-
-        | _ -> assert false
-      end
+    begin
+      let arg_vals, arg_tys, mrepr =
+        List.fold_left
+          (fun (arg_vals, arg_tys, mrepr) arg ->
+             let rep_arg, mrepr = model_repr_of_term arg env mrepr unbounded in
+             rep_arg :: arg_vals,
+             (Expr.type_info arg) :: arg_tys,
+             mrepr
+          )
+          ([], [], mrepr) (List.rev xs)
+      in
+      let ret_rep, mrepr = model_repr_of_term t env mrepr unbounded in
+      match f, arg_vals, ty with
+      | Sy.Name _, [], Ty.Tfarray _ ->
+        begin
+          match Hashtbl.find_opt Cache.arrays_cache t with
+          | Some _ -> acc
+          | None ->
+            (* We have to add an abstract array in case there is no
+               constraint on its values. *)
+            Hashtbl.add Cache.arrays_cache t (Hashtbl.create 17);
+            acc
+        end
+
+      | Sy.Op Sy.Set, _, _ ->
+        (* As arrays are immutable, the result of the set operator is
+           never a user-defined array and has to be ignored. *)
+        acc
+
+      | Sy.Op Sy.Get, [(a, _); (_, i)], _ ->
+        begin
+          match X.term_extract a with
+          | Some ta, true ->
+            let Expr.{ xs = xs_ta; _ } =
+              Expr.term_view ta
+            in
+            assert (xs_ta == []);
+            Cache.store_array_get ta i (ret_rep |> snd);
+            acc
+          | _ ->
+            (* There is no semantic values for arrays, which means an array
+               is directly represented as a term. *)
+            assert false
+        end
 
-    | _ ->
-      if tys == [] then
-        fprofs, ModelMap.add (f, tys, ty) (xs, rep) cprofs, carrays,
-        mrepr
-      else
-        ModelMap.add (f, tys, ty) (xs, rep) fprofs, cprofs, carrays,
-        mrepr
+      | Sy.Name (id, _, _), [], Ty.Trecord trecord ->
```

Comment:
Are we sure nothing special needs to happen for functions returning records?

Reply:
I will check but as it checks there is no argument and a function without argument is the same as a constant in the SMT-LIB standard, it should be ok.

Reply:
What I mean is that we are calling `store_record_sig` for constant records, but not for functions that return records, and I am wondering if we should do something similar for functions that return records.

See e.g. #412 where there is an uninterpreted function that returns a record and where Alt-Ergo gets confused. On a side note, I thought I had fixed the first half of the bug (treating all functions with a single record argument as constructors) but apparently I never made that PR, oops :innocent: . Would be worth checking what happens there with this implementation.